### PR TITLE
[Data Explorer][Discover 2.0] Implement saved search embeddable

### PR DESCRIPTION
**Pull request by ananzh**
Screenshot 2023-09-21 at 12 34 48

## Issue Resolve
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4233

### Overall Functioning

- [x] Dashboard could load a saved search
- [x] saved search should display data grid table
- [x] view single document link is clickable and should see a single doc with two formats (JSON and table)
- [x] view surrounding document link is clickable and should see surrounding documents

### Top Nav
- [x] should display time range of first added saved search 
- [x] change time range can update all saved searches
- [x] filter can update all saved searches
- [x] search query can search all saved searches

### Data grid table
- [x] `filter for value` button in Table should work
- [x] `filter out value` button in Table should work
- [x] should sort
- [x] should be able to move column, except `time` and `_source`
- [x] should be able to remove column. If all columns (except `time`) are removed, should display default `_source`.
- [x] pagination should work

### Context view
- [x] when expand doc, `filter for value` should work
- [x] when expand doc, `filter out value` should work
- [x] when expand doc, toggle column should work
- [x] when expand doc, exist button should work

### Panel Options
- [x] should be able to edit search
- [x] should be able to clone panel
- [x] should be able to edit panel title
- [x] should be able to remove a panel

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/discover/public/embeddable/search_embeddable_factory.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  public createFromSavedObject = async (
+    savedObjectId: string,
+    input: Partial<SearchInput> & { id: string; timeRange: TimeRange },
+    parent?: Container
+  ): Promise<SearchEmbeddable | ErrorEmbeddable> => {
+    const services = getServices();
+    const filterManager = services.filterManager;
+    const url = await services.getSavedSearchUrlById(savedObjectId);
+    const editUrl = services.addBasePath(`/app/discover${url}`);
```
Comment (ashwin-pc) [inline src/plugins/discover/public/embeddable/search_embeddable_factory.tsx:85]:
```suggestion
    const editUrl = services.addBasePath(`/app/data-explorer/discover${url}`);
```

File: src/plugins/discover/public/embeddable/search_embeddable.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/**
+   * 1. We want the osdDocTable__container to scroll only when embedded in an embeddable panel
```
Comment (ashwin-pc) [inline src/plugins/discover/public/embeddable/search_embeddable.scss:2]: Are you sure we want this? The doctable component will already add a scrollbar when the height of the container exceeds the height of the rows combined.

File: src/plugins/discover/public/embeddable/search_embeddable_component.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    onFilter: searchProps.onFilter,
+    onRemoveColumn: searchProps.onRemoveColumn,
+    onSort: searchProps.onSort,
+    rows: searchProps.rows,
+    onSetColumns: searchProps.onSetColumns,
+    sort: searchProps.sort,
+    displayTimeColumn: searchProps.displayTimeColumn,
+    services: searchProps.services,
+    totalHitCount: searchProps.totalHitCount,
+  } as DiscoverEmbeddableProps & { totalHitCount: number };
```
Comment (ashwin-pc) [inline src/plugins/discover/public/embeddable/search_embeddable_component.tsx:41]:
```suggestion
  const discoverEmbeddableProps: DiscoverEmbeddableProps = {
    columns: searchProps.columns,
    indexPattern: searchProps.indexPattern,
    onAddColumn: searchProps.onAddColumn,
    onFilter: searchProps.onFilter,
    onRemoveColumn: searchProps.onRemoveColumn,
    onSort: searchProps.onSort,
    rows: searchProps.rows,
    onSetColumns: searchProps.onSetColumns,
    sort: searchProps.sort,
    displayTimeColumn: searchProps.displayTimeColumn,
    services: searchProps.services,
    totalHitCount: searchProps.totalHitCount,
  };
```
The difference is that this way, typescript ensures that `discoverEmbeddableProps` conforms to `DiscoverEmbeddableProps` and does not forceable change the type of your object to `DiscoverEmbeddableProps`.

Also `DiscoverEmbeddableProps` already has `totalHitCount` as defined on line 20

File: src/plugins/discover/public/embeddable/search_embeddable.tsx [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import './search_embeddable.scss';
+import _ from 'lodash';
```
Comment (ashwin-pc) [inline src/plugins/discover/public/embeddable/search_embeddable.tsx:32]: import only the lodash functions you need for better tree shaking.

**Review by ashwin-pc — COMMENTED** [review]
Love it, great job migrating this so cleanly.

**Review by ashwin-pc — COMMENTED** [review]
Can you add a screenshot to the PR too so that its easy to see what the new change looks like?

**Comment by ananzh** [comment]
> Can you add a screenshot to the PR too so that its easy to see what the new change looks like?

added on the top. also demo it in the developer meeting today.

**Review by ashwin-pc — APPROVED** [review]
Ship it :)

**Pull request merged**